File: lc_py/lcPy/greedy/minSwaps.py
from typing import List


class Solution:
    def minSwaps(self, nums: List[int]) -> int:
        odd_start, even_start, even_pos, odd_pos, even_cnt, odd_cnt = 0, 0, [], [], 0, 0
        for i, x in enumerate(nums):
            if x % 2:
                odd_cnt += 1
                odd_pos.append(i)
            else:
                even_cnt += 1
                even_pos.append(i)
        if abs(even_cnt - odd_cnt) > 1: return -1
        # 偶数开头
        res = inf
        if even_cnt >= odd_cnt:
            target = [i * 2 for i in range(even_cnt)]
            cost = sum(abs(even_pos[i] - target[i]) for i in range(even_cnt))
            res = min(res, cost)
        if odd_cnt >= even_cnt:
            target = [i * 2 for i in range(odd_cnt)]
            cost = sum(abs(odd_pos[i] - target[i]) for i in range(odd_cnt))
            res = min(res, cost)
        return res

File: lc_go/lcGo/greedy/minSwaps.go
package main

import "math"

func minSwaps(nums []int) int {
	var evenPos, oddPos []int
	for i, x := range nums {
		if x%2 == 0 {
			evenPos = append(evenPos, i)
		} else {
			oddPos = append(oddPos, i)
		}
	}

	evenCnt, oddCnt := len(evenPos), len(oddPos)
	if abs(evenCnt-oddCnt) > 1 {
		return -1
	}

	res := math.MaxInt32

	// 偶数开头
	if evenCnt >= oddCnt {
		cost := 0
		for i := 0; i < evenCnt; i++ {
			target := i * 2
			cost += abs(evenPos[i] - target)
		}
		res = int(math.Min(float64(res), float64(cost)))
	}

	// 奇数开头
	if oddCnt >= evenCnt {
		cost := 0
		for i := 0; i < oddCnt; i++ {
			target := i * 2
			cost += abs(oddPos[i] - target)
		}
		res = int(math.Min(float64(res), float64(cost)))
	}

	return res
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func main() {
	nums := []int{1, 2, 3, 4, 5}
	println(minSwaps(nums))
}
